pinot-plugins/pinot-environment/pinot-azure/src/main/java/org/apache/pinot/plugin/provider/AzureEnvironmentProvider.java

```diff
@@ -0,0 +1,159 @@
+        HttpClients.custom().setDefaultRequestConfig(requestConfig).setRetryHandler(httpRequestRetryHandler).build();
+  }
+
+  // Constructor for test purposes.
+  @VisibleForTesting
+  public AzureEnvironmentProvider(int maxRetry, String imdsEndpoint, CloseableHttpClient closeableHttpClient) {
+    _maxRetry = maxRetry;
+    _imdsEndpoint = imdsEndpoint;
+    _closeableHttpClient = Preconditions
+        .checkNotNull(closeableHttpClient, "[AzureEnvironmentProvider]: Closeable Http Client cannot be null");
+  }
+
+  /**
+   *
+   * Utility used to query the azure instance metadata service (Azure IMDS) to fetch the failure domain information,
+   * used at HelixServerStarter startup to update the instance configs.
+   * @return failure domain information
+   */
+  @VisibleForTesting
+  @Override
+  public String getFailureDomain() {
+    final String responsePayload = getAzureInstanceMetadata();
+
+    // For a sample response payload,
+    // check https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service?tabs=linux
+    final ObjectMapper objectMapper = new ObjectMapper();
+    try {
+      final JsonNode jsonNode = objectMapper.readTree(responsePayload);
+      final JsonNode computeNode = jsonNode.path(COMPUTE);
+
+      if (computeNode.isMissingNode()) {
+        throw new RuntimeException(
+            "[AzureEnvironmentProvider]: Compute node is missing in the payload. Cannot retrieve failure domain "
+                + "information");
+      }
+      final JsonNode platformFailureDomainNode = computeNode.path(PLATFORM_FAULT_DOMAIN);
+      if (platformFailureDomainNode.isMissingNode() || !platformFailureDomainNode.isTextual()) {
+        throw new RuntimeException("[AzureEnvironmentProvider]: Json node platformFaultDomain is missing or is invalid."
+            + " No failure domain information retrieved for given server instance");
+      }
+      return platformFailureDomainNode.textValue();
+    } catch (IOException ex) {
+      throw new RuntimeException(String.format(
+          "[AzureEnvironmentProvider]: Errors when parsing response payload from Azure Instance Metadata Service: %s",
+          responsePayload), ex);
+    }
+  }
+
+  // Utility used to construct the HTTP Request and fetch corresponding response entity.
+  @VisibleForTesting
+  private String getAzureInstanceMetadata() {
+    HttpGet httpGet = new HttpGet(_imdsEndpoint);
+    httpGet.setHeader(METADATA, Boolean.TRUE.toString());
+
+    try {
+      final CloseableHttpResponse closeableHttpResponse = _closeableHttpClient.execute(httpGet);
+      if (closeableHttpResponse == null) {
+        throw new RuntimeException("[AzureEnvironmentProvider]: Response is null. Please verify the imds endpoint");
+      }
+      final StatusLine statusLine = closeableHttpResponse.getStatusLine();
+      final int statusCode = statusLine.getStatusCode();
+      if (statusCode != HttpStatus.SC_OK) {
+        final String errorMsg = String
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: statusCode (declared at line 145). This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 147. Consider formatting this data with &#39;%d&#39; instead.